src/components/ColorPicker.tsx

```diff
@@ -42,14 +49,90 @@ const ColorSwatch: FC<{
+      ? addAlphaToHex(rgbToHex(cursorStyle?.backgroundColor))
+      : undefined
+    if (
+      (cursorStyle?.color === undefined && cursorStyle?.backgroundColor === undefined) ||
+      (selectedTextHexColor === '#ccccccff' && selectedBackHexColor === '#333333ff') ||
+      (selectedTextHexColor === addAlphaToHex(rgbToHex(colors.fg)) &&
+        selectedBackHexColor === addAlphaToHex(rgbToHex(colors.bg)) &&
+        !selection.isThought())
+    ) {
+      const colorMatches = currentThoughtValue.match(colorRegex)
+      let matchColor, match
+      // Get the colors and background colors used in current thought's value
+      const fgColors: Set<string> = new Set()
+      if (colorMatches) {
+        colorMatches.forEach(match => fgColors.add(match.slice(7, -1)))
+        matchColor = fgColors.size > 1 ? null : fgColors.values().next().value
+      }
+
+      const bgColors: Set<string> = new Set()
+      while ((match = bgColorRegex.exec(currentThoughtValue)) !== null) if (match[1]) bgColors.add(match[1])
+      const matchBgColor = bgColors.size > 1 ? null : bgColors.values().next().value
+
+      return !!(
+        (textHexColor && textHexColor === (matchColor && addAlphaToHex(rgbToHex(matchColor)))) ||
+        (backHexColor && backHexColor === (matchBgColor && addAlphaToHex(rgbToHex(matchBgColor))))
+      )
+    }
+    return !!(
+      (textHexColor && textHexColor === selectedTextHexColor) ||
+      (backHexColor && backHexColor === selectedBackHexColor)
+    )
+  }, [cursorStyle, backgroundColor, color, currentThoughtValue, colors])
+
   /** Toggles the text color to the clicked swatch. */
   const toggleTextColor = (e: React.MouseEvent | React.TouchEvent) => {
     // stop toolbar button dip
     e.stopPropagation()
+    e.preventDefault()
+
+    if (backgroundColor || color !== 'default') dispatch(formatSelection('foreColor', color || colors.bg))
+    else dispatch(formatSelection('foreColor', colors.fg))
+    // Apply background color to the selection
+    if (backgroundColor && backgroundColor !== colors.bg)
+      dispatch(formatSelection('backColor', backgroundColor === 'inverse' ? colors.fg : backgroundColor))
+    else {
+      dispatch(formatSelection('backColor', colors.bg))
+      const thoughtContentEditable = document.querySelector(`[aria-label="editable-${currentThoughtId}"]`)
+      if (!thoughtContentEditable) return
+      const styledElements = thoughtContentEditable.querySelectorAll('[style]')
+      styledElements?.forEach(el => {
+        const elementColor = el.getAttribute('color')
+        const elementStyle = el.getAttribute('style')
+        if (
+          (elementColor && elementColor !== rgbToHex(colors.bg)) ||
+          (label === 'default' && elementStyle?.includes('background-color'))
+        )
+          el.removeAttribute('style')
+      })
+      dispatch(
+        editThought({ oldValue: currentThoughtValue, newValue: thoughtContentEditable.innerHTML, path: simplePath }),
```

Comment:
Please explain in a comment why the cursor thought is being edited to its innerHTML, as this is not obvious from looking at the code. What problem does it solve? What would occur if this wasn't done?